aiida/repository/control.py

```diff
@@ -0,0 +1,81 @@
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Module for overall repository control commands."""
+# Note: these functions are not methods of `AbstractRepositoryBackend` because they need access to the orm.
+# This is because they have to go through all the nodes to gather the list of keys that AiiDA is keeping
+# track of (since they are descentralized in each node entry).
+# See the get_unreferenced_keyset function
+from aiida.manage.manager import get_manager
+from aiida.orm.implementation import Backend
+
+
+def repository_maintain(full: bool = False, backend: Backend = None, **kwargs) -> dict:
+    """Performs maintenance tasks on the repository."""
+    maintainance_report = {'database': {}, 'repository': {}, 'user_info': ''}
+
+    if backend is None:
+        backend = get_manager().get_backend()
+    repository_backend = backend.get_repository()
+
+    unreferenced_objects = get_unreferenced_keyset(aiida_backend=backend)
+
+    maintainance_report['repository']['unreferenced'] = len(unreferenced_objects)  # type: ignore
+    maintainance_report['repository']['info'] = repository_backend.get_info()  # type: ignore
+
+    repository_backend.delete_objects(list(unreferenced_objects))
+
+    # Perform the maintainance operations in the repository
+    maintainance_report['repository']['maintain'] = repository_backend.maintain(full=full, **kwargs)  # type: ignore
+
+    return maintainance_report
+
+
+def repository_info(statistics: bool = False, backend: Backend = None, **kwargs) -> dict:
+    """Returns relevant information on the repository."""
+    if backend is None:
+        backend = get_manager().get_backend()
+    repository_backend = backend.get_repository()
+    return repository_backend.get_info(statistics)
+
+
+def get_unreferenced_keyset(check_consistency: bool = True, aiida_backend: Backend = None, **kwargs) -> set:
+    """Returns the keyset of objects that exist in the repository but are not tracked by AiiDA.
+
+    This should be all the soft-deleted files.
+    """
+    from aiida import orm
+
+    if aiida_backend is None:
+        aiida_backend = get_manager().get_backend()
+
+    repository_backend = aiida_backend.get_repository()
+
+    keyset_backend = set(repository_backend.list_objects())
+    keyset_aiidadb = set(orm.Node.objects(aiida_backend).iter_repo_keys())
+
+    if check_consistency:
+        if len(keyset_aiidadb - keyset_backend) > 0:
+            raise RuntimeError('Database seems corrupted (some tracked objects are not in the repository). Aborting!')
+
+    return keyset_backend - keyset_aiidadb
+
+
+def get_repository_report(backend: Backend = None, **kwargs) -> dict:
```

Comment:
Is this function really necessary? You are just calling this in the CLI command and are essentially just calling one other function `get_unreferenced_keyset`. The formatting you do would be much better placed in the CLI command than in this function.

Reply:
Currently this is true, but I will add more features related to the repository in an upcoming release (I didn't want to overcrowd this one). From the design document (I highlighted what I would be missing):

> Verdi storage maintain --dry-run
> - Get some heuristics on whether to vacuum PSQL and/or SQLite -> future, open issue
> - Get a count of objects **(and size in MB)** of unreferenced objects (to know you can delete)

Reply:
Also here, I would just have the repo interface expose methods to retrieve this data, and then have the CLI call this. I realize that this means the CLI has to check the type of the repo, but I think that is its purpose. That logic _should_ go there.

Reply:
Mmm, I'm not sure I agree. The  `get_repository_report` method returns information that would help you decide if you need to run maintenance or not (for example, number of unreferenced files that were soft deleted and you could hard delete). This is displayed by the CLI for human users but could be part of a different UI (maybe even for an automated maintenance cron workflow) that would need to do the exact same "backend customization" logic. This is why I separated that from the click function.

Reply:
That's my point though. If you intend this information to be used by other consumers, like scripts, you wouldn't want this information returned in human-readable format, where the important information (the number of unreferenced files) is embedded in a string. The interface _should_ have an easy method to retrieve this data, but it should not be bothered how it is presented to a user; that is the CLI's job.